modules/plugin/src/main/scala/com/fortysevendeg/hood/plugin/SbtHoodPlugin.scala

```diff
@@ -0,0 +1,137 @@
+
+object SbtHoodPlugin extends AutoPlugin with SbtHoodDefaultSettings with SbtHoodKeys {
+
+  override def projectSettings: Seq[Def.Setting[_]] = defaultSettings
+  override val trigger: PluginTrigger               = noTrigger
+
+  def compareBenchmarksTask: Def.Initialize[Task[Unit]] = Def.task {
+
+    implicit val logger = Slf4jLogger.getLogger[IO]
+
+    benchmarkTask(
+      previousBenchmarkPath.value,
+      currentBenchmarkPath.value,
+      keyColumnName.value,
+      compareColumnName.value,
+      thresholdColumnName.value,
+      modeColumnName.value,
+      unitsColumnName.value,
+      generalThreshold.value,
+      benchmarkThreshold.value
+    ).leftFlatMap(e =>
+        EitherT.left[List[BenchmarkComparisonResult]](logger.error(s"There was an error: $e")))
+      .value
+      .unsafeRunSync()
+
+    ()
+  }
+
+  def benchmarkTask[F[_]](
+      previousPath: File,
+      currentPath: File,
+      keyColumnName: String,
+      compareColumnName: String,
+      thresholdColumnName: String,
+      modeColumnName: String,
+      unitsColumnName: String,
+      generalThreshold: Option[Double],
+      benchmarkThreshold: Map[String, Double])(
+      implicit L: Logger[F],
+      S: Sync[F],
+      C: Console[F]): EitherT[F, HoodError, List[BenchmarkComparisonResult]] = {
+
+    val csvService = CsvService.build[F]
+
+    val parseFunction: File => F[Either[HoodError, List[Benchmark]]] = csvService.parseBenchmark(
+      keyColumnName,
+      modeColumnName,
+      compareColumnName,
+      thresholdColumnName,
+      unitsColumnName)
+
+    for {
+      previousBenchmarks <- EitherT(parseFunction(previousPath)).map(buildBenchmarkMap)
+      currentBenchmarks  <- EitherT(parseFunction(currentPath)).map(buildBenchmarkMap)
+      result <- EitherT.right[HoodError](
+        performBenchmarkComparison[F](
+          currentBenchmarks,
+          previousBenchmarks,
+          generalThreshold,
+          benchmarkThreshold))
+      outputMessage = benchmarkOutput(result, previousPath.getName, currentPath.getName)
+      _ <- EitherT.right(C.putStrLn(outputMessage))
+    } yield result
+
+  }
+
+  private[this] def buildBenchmarkMap(benchmarks: List[Benchmark]): Map[String, Benchmark] =
+    benchmarks.map(b => (b.benchmark, b)).toMap
+
+  private[this] def performBenchmarkComparison[F[_]](
+      currentBenchmarks: Map[String, Benchmark],
+      previousBenchmarks: Map[String, Benchmark],
+      generalThreshold: Option[Double],
+      thresholdMap: Map[String, Double])(
+      implicit L: Logger[F],
+      S: Sync[F]): F[List[BenchmarkComparisonResult]] =
+    S.delay(previousBenchmarks.map {
+      case (benchmarkKey, previous) =>
+        val threshold =
+          thresholdMap
+            .get(benchmarkKey)
+            .fold(generalThreshold.getOrElse(previous.primaryMetric.scoreError))(identity)
+
+        currentBenchmarks
+          .get(benchmarkKey)
+          .fold({
+            L.error(
```

Comment:
Potentially not executed, since it's inside a `delay` and is a side-effect

Reply:
Good catch! Please take a look at the fix I've submitted when you have a chance ;)